lib/import_export/base_exporter.rb

```diff
@@ -0,0 +1,161 @@
+module ImportExport
+  class BaseExporter
+
+    attr_reader :export_data, :categories, :topics
+
+    CATEGORY_ATTRS = [:id, :name, :color, :created_at, :user_id, :slug, :description, :text_color,
+                      :auto_close_hours, :parent_category_id, :auto_close_based_on_last_post,
+                      :topic_template, :suppress_from_homepage, :all_topics_wiki, :permissions_params]
+
+    GROUP_ATTRS = [ :id, :name, :created_at, :mentionable_level, :messageable_level, :visibility_level,
+                    :automatic_membership_email_domains, :automatic_membership_retroactive,
+                    :primary_group, :title, :grant_trust_level, :incoming_email]
+
+    USER_ATTRS = [:id, :email, :username, :name, :created_at, :trust_level, :active, :last_emailed_at]
+
+    TOPIC_ATTRS = [:id, :title, :created_at, :views, :category_id, :closed, :archived, :archetype]
+
+    POST_ATTRS = [:id, :user_id, :post_number, :raw, :created_at, :reply_to_post_number, :hidden,
+                  :hidden_reason_id, :wiki]
+
+    def categories
+      @categories ||= Category.all.to_a
+    end
+
+    def export_categories
+      data = []
+
+      categories.each do |cat|
+        data << CATEGORY_ATTRS.inject({}) { |h, a| h[a] = cat.send(a); h }
+      end
+
+      data
+    end
+
+    def export_categories!
+      @export_data[:categories] = export_categories
+
+      self
+    end
+
+    # export groups that are mentioned in category permissions
+    def export_category_groups
+      groups = []
+      group_names = []
+      auto_group_names = Group::AUTO_GROUPS.keys.map(&:to_s)
+
+      @export_data[:categories].each do |c|
+        c[:permissions_params].each do |group_name, _|
+          group_names << group_name unless auto_group_names.include?(group_name.to_s)
+        end
+      end
+
+      group_names.uniq!
+      return [] if group_names.empty?
+
+      Group.where(name: group_names).find_each do |group|
+        attrs = GROUP_ATTRS.inject({}) { |h, a| h[a] = group.send(a); h }
+        attrs[:user_ids] = group.users.pluck(:id)
+        groups << attrs
+      end
+
+      groups
+    end
+
+    def export_category_groups!
+      @export_data[:groups] = export_category_groups
+
+      self
+    end
+
+    def export_group_users
+      user_ids = []
+
+      @export_data[:groups].each do |g|
+        user_ids += g[:user_ids]
+      end
+
+      user_ids.uniq!
+      return [] if user_ids.empty?
+
+      users = User.where(id: user_ids)
+      export_users(users)
+    end
+
+    def export_group_users!
+      @export_data[:users] = export_group_users
+
+      self
+    end
+
+    def export_topics
+      data = []
+
+      topics.each do |topic|
+        puts topic.title
```

Comment:
stray puts?

Reply:
Not for log purpose actually. Old exporter itself having this code to show the topic title **in console window** which is exporting while export process.